inc/classes/CDN/CDN.php

```diff
@@ -0,0 +1,297 @@
+		 * Filter all CDN URLs.
+		 *
+		 * @since 2.7
+		 *
+		 * @param array $hosts List of CDN URLs.
+		 */
+		$hosts = (array) apply_filters( 'rocket_cdn_cnames', $hosts );
+		$hosts = array_filter( $hosts );
+		$hosts = array_flip( array_flip( $hosts ) );
+		$hosts = array_values( $hosts );
+
+		return $hosts;
+	}
+
+	/**
+	 * Gets the base URL for the website
+	 *
+	 * @since 3.4
+	 * @author Remy Perona
+	 *
+	 * @return string
+	 */
+	private function get_base_url() {
+		$home = get_option( 'home' );
+
+		return substr( $home, strpos( $home, '//' ) );
+	}
+
+	/**
+	 * Gets the allowed paths as a regex pattern for the CDN rewrite
+	 *
+	 * @since 3.4
+	 * @author Remy Perona
+	 *
+	 * @return string
+	 */
+	private function get_allowed_paths() {
+		$wp_content_dirname = ltrim( wp_parse_url( content_url(), PHP_URL_PATH ), '/' );
+
+		$upload_dirname = '';
+		$uploads_info   = wp_upload_dir();
+
+		if ( ! empty( $uploads_info['baseurl'] ) ) {
+			$upload_dirname = '|' . ltrim( trailingslashit( wp_parse_url( $uploads_info['baseurl'], PHP_URL_PATH ) ), '/' );
+		}
+
+		return $wp_content_dirname . $upload_dirname . '|wp-includes';
+	}
+
+	/**
+	 * Checks if the provided URL can be rewritten with the CDN URL
+	 *
+	 * @since 3.4
+	 * @author Remy Perona
+	 *
+	 * @param string $url URL to check.
+	 * @return boolean
+	 */
+	private function is_excluded( $url ) {
+		if ( 'php' === pathinfo( strtok( $url, '?' ), PATHINFO_EXTENSION ) ) {
+			return true;
+		}
+
+		$path = wp_parse_url( $url, PHP_URL_PATH );
+
+		if ( ! isset( $path ) ) {
+			return true;
+		}
+
+		if ( '/' === $path ) {
+			return true;
+		}
+
+		if ( preg_match( '#^(' . $this->get_excluded_files() . ')$#', $path ) ) {
+			return true;
+		}
+
+		return false;
+	}
+
+	/**
+	 * Gets the CDN zones for the provided URL
+	 *
+	 * @since 3.4
+	 * @author Remy Perona
+	 *
+	 * @param string $url URL to check.
+	 * @return array
+	 */
+	private function get_zones_for_url( $url ) {
+		$zones = [ 'all' ];
+
+		$ext = pathinfo( $url, PATHINFO_EXTENSION );
+
+		$image_types = [
+			'jpg',
+			'jpeg',
+			'jpe',
+			'png',
+			'gif',
+			'webp',
+			'bmp',
+			'tiff',
+			'svg',
+		];
+
+		if ( 'css' === $ext || 'js' === $ext ) {
+			$zones[] = 'css_and_js';
+		}
+
+		if ( 'css' === $ext ) {
+			$zones[] = 'css';
+		}
+
+		if ( 'js' === $ext ) {
+			$zones[] = 'js';
+		}
+
+		if ( in_array( $ext, $image_types, true ) ) {
+			$zones[] = 'images';
+		}
+
+		return $zones;
+	}
+
+	/**
+	 * Get all files we don't allow to get in CDN.
+	 *
+	 * @since 2.5
+	 *
+	 * @return string A pipe-separated list of excluded files.
+	 */
+	private function get_excluded_files() {
+		$files = $this->options->get( 'cdn_reject_files', [] );
+
+		/**
+			* Filter the excluded files.
+			*
+			* @since 2.5
+			*
+			* @param array $files List of excluded files.
+		*/
+		$files = (array) apply_filters( 'rocket_cdn_reject_files', $files );
+		$files = array_filter( $files );
+		$files = array_flip( array_flip( $files ) );
```

Comment:
It should be `preg_quote()`d, maybe by passing the delimiter as method’s parameter.